src/ReachSets/DiscretePost/DecomposedDiscretePost.jl

```diff
@@ -0,0 +1,144 @@
+export DecomposedDiscretePost
+
+"""
+    DecomposedDiscretePost <: DiscretePost
+
+Textbook implementation of a discrete post operator, but with lazy decomposed intersections.
+
+### Fields
+
+- `options` -- an `Options` structure that holds the algorithm-specific options
+
+### Algorithm
+
+The algorithm is based on [Flowpipe-Guard Intersection for Reachability
+Computations with Support Functions](http://spaceex.imag.fr/sites/default/files/frehser_adhs2012.pdf).
+"""
+struct DecomposedDiscretePost <: DiscretePost
+    options::Options
+
+    function DecomposedDiscretePost(𝑂::Options)
+        𝑂copy = copy(𝑂)
+        # TODO: Check why it takes always default value for convex_hull
+        check_aliases_and_add_default_value!(𝑂.dict, 𝑂copy.dict, [:overapproximation], Hyperrectangle)
+        check_aliases_and_add_default_value!(𝑂.dict, 𝑂copy.dict, [:out_vars], Vector{Int}())
+
+        return new(𝑂copy)
+    end
+end
+
+# convenience constructor from pairs of symbols
+DecomposedDiscretePost(𝑂::Pair{Symbol,<:Any}...) = DecomposedDiscretePost(Options(Dict{Symbol,Any}(𝑂)))
+
+# default options for the DecomposedDiscretePost discrete post operator
+DecomposedDiscretePost() = DecomposedDiscretePost(Options())
+
+init(𝒫::DecomposedDiscretePost, 𝒮::AbstractSystem, 𝑂::Options) = init!(𝒫, 𝒮, copy(𝑂))
+
+# TODO: use 𝑂 only?
+function init!(𝒫::DecomposedDiscretePost, 𝒮::AbstractSystem, 𝑂::Options)
+    𝑂[:n] = statedim(𝒮, 1)
+
+    # solver-specific options (adds default values for unspecified options)
+    𝑂out = validate_solver_options_and_add_default_values!(𝑂)
+
+    return 𝑂out
+end
+
+function tube⋂inv!(𝒫::DecomposedDiscretePost,
+                   reach_tube::Vector{<:ReachSet{<:LazySet, N}},
+                   invariant,
+                   Rsets,
+                   start_interval
+                  ) where {N}
+
+    dirs = 𝒫.options[:overapproximation]
+
+    # counts the number of sets R⋂I added to Rsets
+    count = 0
+    @inbounds for reach_set in reach_tube
+        push!(Rsets, ReachSet{LazySet{N}, N}(reach_set.X,
+            reach_set.t_start + start_interval[1],
+            reach_set.t_end + start_interval[2]))
+        count = count + 1
+    end
+
+    return count
+end
+
+function post(𝒫::DecomposedDiscretePost,
+              HS::HybridSystem,
+              waiting_list::Vector{Tuple{Int, ReachSet{LazySet{N}, N}, Int}},
+              passed_list,
+              source_loc_id,
+              tube⋂inv,
+              count_Rsets,
+              jumps,
+              options
+             ) where {N}
+    jumps += 1
+    oa = 𝒫.options[:overapproximation]
+    temp_vars = 𝒫.options[:temp_vars]
```

Comment:
`temp_vars` is only used in `lenght(temp_vars)`, so you can instead do
```suggestion
    n_lowdim = length(𝒫.options[:temp_vars])
```
Should we better check that `n_lowdim < n`? But it is not clear where to get `n` from (maybe from `options[:n]`?).

Reply:
I didn't want to produce too many `artificial` options. Should I add option for `n`?

Reply:
No, but did you try `options[:n]`?

Reply:
Nope, too many options, I even didn't know that this option exist:(

Reply:
And actually it doesn't work in the case of DiscretePostOperator. We can update this option inside solve, but I am not sure wether it is a good idea.

Reply:
See [here](https://github.com/JuliaReach/Reachability.jl/blob/master/src/ReachSets/DiscretePost/LazyDiscretePost.jl#L76)

Reply:
Hm, the problem is that the algorithm is broken for `n_lowdim == n` because of this check:
```julia
if (dim(reach_set.X) == n_lowdim)
    continue
end
```
So I think we should find a way of verifying this condition or avoid this check.

> See here

This line (`𝑂[:n] = statedim(𝒮, 1)`) exists in `DecomposedDiscretePost.jl` as well, so I would expect it to work.

Reply:
> This line (𝑂[:n] = statedim(𝒮, 1)) exists in DecomposedDiscretePost.jl as well, so I would expect it to work.

I know, `𝒮` is an `AbstractSystem` which is high-dimensional. Therefore, the result of this line should be the dimension of high-dimensional set.

Reply:
But this is what we want, right?

Reply:
We have to intersect with guards and target invariants if:
- `reach_set` is computed in high-dimension, instead of low dimension (`n_lowdim < n`)
-  if `n_lowdim == n`

so it should something like:
```
if (dim(reach_set.X) == n_lowdim && n_lowdim < n)
         continue
end
# check intersection with guard
R⋂G = Intersection(reach_set.X, guard)
if isempty(R⋂G)
        continue
end
R⋂G = overapproximate(R⋂G, CartesianProductArray, oa)
```

Reply:
True!